Add agent type

rf-ivtdai opened the pull request:
client-side update to accommodate adding agent_type to agent attributes

Ruishenl commented on invertedai/common.py:
```diff
@@ -141,22 +141,35 @@ class AgentAttributes(BaseModel):
     #: Distance from the agent's center to its rear axis in meters. Determines motion constraints.
-    rear_axis_offset: float
+    rear_axis_offset: Optional[float] = None
+    agent_type: Optional[str] = 'car'  #: Type of agent, default to car, can also be pedestrian
```
Can you make the type here the Enum instead of str ?

adscib replied:
I'm actually not sure that an enum is preferable. It would force people to update the package whenever we introduce additional types.

Ruishenl replied:
If we are using string enums it will be automatically casted as string and when we introduce new types we just append more values. Plus it is  using enums for validation already on client side

rf-ivtdai replied:
keeping string as per our conversation.

adscib commented on invertedai/common.py:
```diff
@@ -49,7 +49,7 @@ def fromlist(cls, l):
 
     def __sub__(self, other):
-        return math.sqrt((abs(self.x - other.x)**2) + (abs(self.y - other.y)**2))
+        return math.sqrt((abs(self.x - other.x) ** 2) + (abs(self.y - other.y) ** 2))
```
`abs` is redundant when taking a square.

adscib commented on invertedai/common.py:
```diff
@@ -141,22 +141,35 @@ class AgentAttributes(BaseModel):
+            length, width, rear_axis_offset, agent_type = l
+            return cls(length=length, width=width, rear_axis_offset=rear_axis_offset, agent_type=agent_type)
+        else:
+            agent_type = l[0]
```
If you're asserting that `l` only has one element, `agent_type, = l` would be preferable. If it can have more, I'm a little worried about how you know that the agent type comes first.

Ruishenl replied:
It's probably safer to do agent_type = l[-1]

adscib commented on invertedai/api/validations.py:
```diff
@@ -0,0 +1,62 @@
+
+def validate_attributes_without_history(agent_attributes: List[AgentAttributes]):
+    if agent_attributes is not None:
+        if not all([len(agent.tolist()) == 1 and type(agent.agent_type) is str for agent in agent_attributes]):
```
Is it actually illegal to use `agent_type=None` here?

adscib commented on invertedai/api/validations.py:
```diff
@@ -0,0 +1,62 @@
+
+def validate_attributes_with_history(states_history: List[List[AgentState]], agent_attributes: List[AgentAttributes],
+                                     agent_count: int):
+    if agent_attributes is None:
```
This is not allowed under the type signature.

adscib commented on invertedai/api/validations.py:
```diff
@@ -0,0 +1,62 @@
+            )
+        if len(states_history[0]) > agent_count:
+            raise InvalidInput(
+                f"num_agents_to_spawn ({agent_count}) must be larger than the number of conditional agents"
```
I guess "not smaller"?

adscib commented on invertedai/api/validations.py:
```diff
@@ -0,0 +1,62 @@
+        raise InvalidInputType(
+            "agent attributes: need to specify length, width, rear_axis_offsets given full state history.")
+    else:
+        if len(states_history[0]) != len(agent_attributes):
```
Should we also check that all lists in states_history have the same length?

adscib commented on invertedai/api/validations.py:
```diff
@@ -0,0 +1,62 @@
+def validate_drive_flows(agent_states: List[AgentState],
+                         agent_attributes: List[AgentAttributes],
+                         recurrent_states: List[RecurrentState]):
+    if len(agent_states) != len(agent_attributes):
+        raise InvalidInput("Incompatible Number of Agents in either 'agent_states' or 'agent_attributes'.")
+    if len(agent_states) != len(recurrent_states):
+        raise InvalidInput("Incompatible Number of Agents in either 'agent_states' or 'recurrent_states'.")
+    if not all([agent.agent_type == AgentType.car.value for agent in agent_attributes]):
+        raise InvalidInput("unsupported agent type in agents_attributes.")
```
Here and throughout, the more information you can include in the error message, the better. For example, you can specify the lengths of both lists, and the first unsupported agent type found.

adscib reviewed (changes requested):
Generally looks good. I left some comments to make sure that the validators match the intended semantics.

We should also update documentation throughout. At the very least, update the user guide to indicate that we support pedestrians, and the main docstring for DRIVE to warn that it doesn't currently support pedestrians, but will in the future.

Ruishenl commented on invertedai/api/validations.py:
```diff
@@ -0,0 +1,62 @@
+            )
+        if not all([all_float_fields_valid(agent) for agent in agent_attributes]):
+            raise InvalidInput(
+                "agent attributes: need to specify length, width, rear_axis_offsets in valid range,"
```
We should mention pedestrian doesn't have lr in the error messages here

Ruishenl commented on invertedai/api/validations.py:
```diff
@@ -0,0 +1,62 @@
+    elif agent.agent_type == AgentType.pedestrian.value:
+        return all([type(field) is float and field >= 0 for field in agent.tolist()[:2]])
+    else:
+        return False
```
Maybe we should raise a type not supported exception here ?

rf-ivtdai commented:
I have addressed the comments and documented the new features in API and class docstrings. I am a bit unsure as to whether we want to keep the client-side request validations for initialize and drive. After testing, the current checks are all handled gracefully by the server. Replicating the same logic here is redundant and error-prone.

adscib commented on invertedai/common.py:
```diff
@@ -141,22 +141,34 @@ class AgentAttributes(BaseModel):
     #: Distance from the agent's center to its rear axis in meters. Determines motion constraints.
-    rear_axis_offset: float
+    rear_axis_offset: Optional[float] = None
+    agent_type: Optional[str] = 'car'  #: Type of agent, default to car, can also be pedestrian
```
```suggestion
    agent_type: Optional[str] = 'car'  #: Valid types are those in `AgentType`, but we use `str` here for extensibility.
```
I actually liked having an `AgentType` enum, even though we use `str` here. We need to have an authoritative list of valid agent types somewhere, and it only makes sense that we would expose it programmatically. Let's just be clear that this enum is extensible and we expect it to grow in future releases.

adscib reviewed (commented):
I updated the docstring, but please make sure it is accurate.

rf-ivtdai commented:
To better measure alignments between client and server, i'll add regression tests that mirrors the server integration tests for initialize and drive

KieranRatcliffeInvertedAI commented:
Changes to the C++ SDK are complete, @adscib is there anything else that needs to be changed before we can merge?

rf-ivtdai commented:
Regression tests added:
initialize: 10 positive flow - expect 200
                  5 negative flow - expect 400
drive: 4 positive flow - expect 200
drive: 1 negative flow - expect 400
These are all relevant to add_agent_types feature, all passing against local server

KieranRatcliffeInvertedAI commented:
@adscib I know @rf-ivtdai has been talking to me about adding regression tests which he seems to have added some on the python SDK. Should we add some to C++ SDK or should that be a different PR?

adscib commented:
It would be nice to add some tests, but I don't want to delay this release, so a separate PR please. Are we waiting for anything here, or can we release already?

rf-ivtdai commented:
We are ready to release!
